### PR TITLE
Doc added for constexpr support

### PR DESCRIPTION
**Pull request by rezarastak**
Follow up after #10047

### CODE REVIEW

File: doc/doxygen/headers/c++.h [hunk 1]
```diff
@@ -54,6 +54,52 @@
  * the idiomatic use of range-based for loops. Examples are
  * GeometryInfo::face_indices(), GeometryInfo::vertex_indices(),
  * FEValuesBase::quadrature_point_indices(), among many others.
+ *
+ * C++11 also introduces the concept of
+ * [constexpr](https://en.cppreference.com/w/cpp/language/constexpr)
```
Comment (kronbichler) [inline doc/doxygen/headers/c++.h:59]: Have you checked that doxygen can render this markup link style appropriately?

File: doc/doxygen/headers/c++.h [hunk 2]
```diff
@@ -54,6 +54,52 @@
...
+ * require a compiler with C++14 capability. As such, this function is
+ * internally declared as:
+ * @code
+ * template <int dim, typename Number>
+ * DEAL_II_CONSTEXPR Number determinant(const Tensor<2, dim, Number> &t);
+ * @endcode
+ * The macro @ref DEAL_II_CONSTEXPR simplifies to `constexpr` if a C++14-capable
+ * compiler is available. Otherwise, for old compilers, it ignores
+ * DEAL_II_CONSTEXPR altogether.
+ * Therefore, in newer compilers, the users of dealii can write
```
Comment (kronbichler) [inline doc/doxygen/headers/c++.h:95]:
```suggestion
 * Therefore, with newer compilers, the user can write
```

File: doc/doxygen/headers/c++.h (hunk 1 above)
Comment (rezarastak) [reply doc/doxygen/headers/c++.h:59]: Yes I checked the html and all of the links inside it.

File: doc/doxygen/headers/c++.h (hunk 2 above)
Comment (rezarastak) [reply doc/doxygen/headers/c++.h]: fixed

File: doc/doxygen/headers/c++.h [hunk 3]
```diff
@@ -54,6 +54,52 @@
...
+ * As an example, since the constructor Tensor::Tensor(const array_type &) is
+ * `constexpr`, we can initialize a tensor with an array during compile time
+ * as:
+ * @code
+ * constexpr double[2][2] entries = {{1., 0.}, {0., 1.}};
+ * constexpr Tensor<2, 2> A(entries);
+ * @endcode
+ * Here, the contents of A are not stored on the stack. Rather, they are
+ * initialized during compile time and inserted into the `.data` portion
+ * of the executable program. The program can use these values in runtime
```
Comment (bangerth) [inline doc/doxygen/headers/c++.h:79]:
```suggestion
 * of the executable program. The program can use these values at runtime
```

File: doc/doxygen/headers/c++.h [hunk 4]
```diff
@@ -54,6 +54,52 @@
  * the idiomatic use of range-based for loops. Examples are
  * GeometryInfo::face_indices(), GeometryInfo::vertex_indices(),
  * FEValuesBase::quadrature_point_indices(), among many others.
+ *
+ * C++11 also introduces the concept of
+ * [constexpr](https://en.cppreference.com/w/cpp/language/constexpr)
+ * variables and function. The variables defined as `constexpr` are constant
+ * values that are computed during the compilation of the program and therefore
+ * have zero runtime cost associated to their initialization. Additionally,
```
Comment (bangerth) [inline doc/doxygen/headers/c++.h:62]:
```suggestion
 * have zero runtime cost associated with their initialization. Additionally,
```

**Comment by rezarastak** [comment]
This PR was part of #8003

**Pull request merged**